### PR TITLE
[Graphbolt] Add load_feature_stores

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/feature_store.py [hunk 1]
```diff
@@ -134,3 +138,81 @@ def update(self, value: torch.Tensor, ids: torch.Tensor = None):
                 f"but got {ids.shape[0]} and {value.shape[0]}."
             )
             self._tensor[ids] = value
+
+
+def load_feature_stores(root_dir: str, feat_desc: dict, in_memory: bool):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/feature_store.py:143]: The metadata in canonical data format should have the information per feature store e.g.
edge_feature: {
  "a", "do", "b": {
    "format": "npy"
    "path": "blabla"
    "in_memory": no
  }
  "a", "did", "c": {
    "format": "csv"
    "path": "blabla"
    "in_memory": yes
  }
}

@Rhett-Ying

Comment (frozenbugs) [reply python/dgl/graphbolt/feature_store.py:143]: For this load_feature_stores, the input should just be the metadata.

Comment (Rhett-Ying) [reply python/dgl/graphbolt/feature_store.py:143]: As discussed offline, `gb.Dataset` will pass in a metadata class which has `format`, `path`, `in_memory` attributes.
```
class OnDiskFeatureData(pydantic_yaml.YamlModel):
    format: Enum[numpy]
    path: str
    in_memory: bool = False
```

File: python/dgl/graphbolt/feature_store.py [hunk 2]
```diff
@@ -134,3 +140,73 @@ def update(self, value: torch.Tensor, ids: torch.Tensor = None):
                 f"but got {ids.shape[0]} and {value.shape[0]}."
             )
             self._tensor[ids] = value
+
+
+class OnDiskFeatureData(pydantic.BaseModel):
+    r"""The description of an on-disk feature."""
+    name: str
+    format: str
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/feature_store.py:148]: try with enum like this https://github.com/dmlc/dgl/blob/8adb53bb1c7cef11efb3d769455521757e583f97/python/dgl/graphbolt/dataset.py#L69 /

File: python/dgl/graphbolt/feature_store.py [hunk 3]
```diff
@@ -134,3 +140,73 @@ def update(self, value: torch.Tensor, ids: torch.Tensor = None):
                 f"but got {ids.shape[0]} and {value.shape[0]}."
             )
             self._tensor[ids] = value
+
+
+class OnDiskFeatureData(pydantic.BaseModel):
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/feature_store.py:154]: @Rhett-Ying Maybe consider moving all "OnDisk" implementation to a single file?

**Comment by czkkkkkk** [comment]
@dgl-bot

File: python/dgl/graphbolt/feature_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/graphbolt/feature_store.py:154]: I will handle such duplication when init feature storage for dataset.

File: tests/python/pytorch/graphbolt/test_feature_store.py [hunk 4]
```diff
@@ -57,3 +57,42 @@ def test_torch_based_feature_store(in_memory):
         # it before closing the temporary directory.
         a = b = None
         feat_store_a = feat_store_b = None
+
+
+def write_tensor_to_disk(dir, name, t, fmt="pt"):
+    if fmt == "pt":
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_feature_store.py:63]: use `torch`/`numpy` to be consistent with the enum we've defined?

**Review by Rhett-Ying — APPROVED** [review]

File: tests/python/pytorch/graphbolt/test_feature_store.py (hunk 4 above)
Comment (Rhett-Ying) [reply tests/python/pytorch/graphbolt/test_feature_store.py:63]: I will fix it in another PR

**Comment by Rhett-Ying** [comment]
CI failed due to dist test timeout. it's ok to merge this PR.

**Pull request merged**